proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+
+Examples
+--------
+
+* 	One of the simplest examples is this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x) = "RULE FIRED" #-}
+
+	It would match expressions like:
+	::
+
+		foo (\x -> x * 2 + x)
+
+* 	The higher order pattern may involve multiple locally bound variables, e.g.:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y z -> f x y z) = "RULE FIRED" #-}
+
+	Which would match:
+	::
+
+		foo (\x y z -> x * y + z)
+
+	But not every variable has to occur in the match. It would also match this expression where ``y`` does not occur:
+	::
+
+		foo (\x y z -> x * 2 + z)
+
+* 	Locally bound variables may only occur once.
+	Consider the following rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x -> f x x) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x -> x * 2 + x)
+
+	But it does contain the valid subrule ``f x``, so it would match:
+	::
+
+		foo (\x -> (bar x . baz) x)
+
+* 	Similarly if the template variable ``f`` is applied to non-variable arguments then it only matches a literal application.
+	Consider this rule:
+	::
+
+		{-# RULES "foo" forall f. foo (\x y -> f x 2 y) = "RULE FIRED" #-}
+
+	This would **not** match:
+	::
+
+		foo (\x y -> x * 2 + y)`
+
+	But again it does contain the higher order pattern ``f x``, so it would match:
+	::
+
+		foo (\x y -> (bar x . baz) 2 y)
+
+Effect and Interactions
+-----------------------
+
+The main effect of this proposal is that rewrite rules involving higher order patterns now match more expressions.
+But the additional matches are guaranteed to be beta equivalent, so this change does not cause existing rules to become semantically incorrect.
+
+The only contentious interactions could occur due to rules that now overlap under the new rules, for example:
+::
+
+	{-# RULES
+	"foo->bar"  forall f x.  foo x (\y. f y) = bar x f
+	"foo->baz"  forall   x.  foo x (\y. y * 2 + y) = baz x
+	#-}
+
+Previously, the rule ``"foo->baz"`` would always fire when encountering the expression ``foo x (\y. y * 2 + y)``, but now the rule ``"foo->bar"`` also matches.
+However, we do not expect that this occurs in practice.
+
+
+Costs and Drawbacks
+-------------------
+
+1. 	The changes required for this proposal are small (the core of the change is an addition of just 22 lines of code).
+	Small changes can add up, but we think the benefits far outweigh this cost in this case.
+
+2. 	This proposal causes a silent change of behaviour of existing code.
+	It is possible to come up with an artificial system of rewrite rules that produces suboptimal results due to this change.
+	We do not expect this to happen in practice.
```

Comment:
Point to the example. Is there one? I don't recall any, at least not where we let a *different* rewrite rule win vs. the status quo. Subsitutions might be different, too, but still equal modulo eta.

Reply:
Actually I'm not sure what this paragraph means.  @noughtmare ?

Reply:
I meant the example from the effects and interactions section right before it.

Reply:
Ah, I see that is not really a problem due to GHC picking the most specific rule. 

I still think it should be possible to construct a set of rules that fires differently.